File: source/cpp/6. Zigzag Conversion.cpp
#include <numeric>
#include <string>
#include <string_view>
#include <vector>


class Solution {
public:
    std::string convert(std::string_view s, const int& numRows) const {
        if (numRows == 1) return std::string(s);

        std::vector<std::string> rows(numRows);
        auto row = 0;
        bool down = true;

        for (const auto& c : s) {
            rows[row] += c;
            
            if (row == 0) down = true;
            else if (row == numRows - 1) down = false;

            row += down ? 1 : -1;
        }

        return std::accumulate(rows.begin(), rows.end(), std::string());
    }
};

File: source/python/6. Zigzag Conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s

        rows = [''] * numRows
        row = 0
        down = True

        for c in s:
            rows[row] += c
            if row == 0:
                down = True
            elif row == numRows - 1:
                down = False

            row += 1 if down else -1

        return ''.join(rows)


def test():
    s = Solution()

    assert s.convert("PAYPALISHIRING", 3) == "PAHNAPLSIIGYIR"
    assert s.convert("PAYPALISHIRING", 4) == "PINALSIGYAHRPI"
    assert s.convert("A", 1) == "A"

    print("All tests passed!")


if __name__ == "__main__":
    test()
